lib/Parser/Parse.cpp

```diff
@@ -8061,7 +8059,6 @@ ParseNodeClass * Parser::ParseClassDecl(BOOL isDeclaration, LPCOLESTR pNameHint,
             pnodeConstructor->hintLength = constructorNameLength;
             pnodeConstructor->hintOffset = constructorShortNameHintOffset;
             pnodeConstructor->pid = pnodeName && pnodeName->pid ? pnodeName->pid : wellKnownPropertyPids.constructor;
-            pnodeConstructor->SetHasNonThisStmt();
```

Comment:
I think this is because we'll defer-parse the constructor function now and when we do, it might not have a non-this statement so this flag won't be set on the defer-parsed function. This flag affects how we store things into slots, so we'll get confused loading symbols from the restored ScopeInfo. I don't think this flag is actually required in the first place but didn't cause a problem without defer parsing.

Reply:
"This flag affects how we store thing into slots." How?

Reply:
Ah sorry that was not the issue. I took a look back. What this flag affects is `JavascriptOperators::UpdateNewScObjectCache`. The defer parsed function might not have this flag set which causes us to go down a different path in that function and we'll hit an assert later. 

I removed this flag and nothing breaks so it wasn't clear to me why we wanted to do this in the first place. If there's a good reason, I guess we can set the flag up when we defer parse a function and we know it's a class constructor.

Reply:
The flag is there to indicate whether the constructor cache can safely predict the type of the constructed object on exit from the function. We say it can if all instructions are of the form "this.x = [simple expr with no side effects]". What is the assert we hit if the flag is set?

Reply:
I believe the assert we hit was in `ConstructorCache::TryUpdateAfterConstructor` and I think it was this assert `Assert(this->content.ctorHasNoExplicitReturnValue);`. 

That seems to make sense why we forced this flag for all class constructors. A base class constructor can return an object constructed from an arbitrary new.target as part of a super chain. 

I wonder how this works for regular constructors, though. A class can derive from any script function or builtin. Then that base function would return an object constructed from the derived class via new.target, right?

Reply:
Is new.target important here? Do we somehow use new.target to determine the type we should create for the new object? We may, but we may also just create a type that's appropriate for the base class and let the type evolve as the derived constructor(s) add their own properties. 

And can you tell what the assert is about? Is it just reporting that we could have used the constructor cache but didn't? If so, how does it know, and what does it have to do with whether there's an explicit return?

Reply:
I'd guess a return statement is a non-this statement so we're just asserting that the constructor which says it doesn't have any non-this statement also doesn't have an explicit return statement. 

We use new.target to construct the object we'll bind to `this`. Any function can end up being new.target but I don't think we do anything otherwise special about using new.target to figure the type of the instance.

Reply:
I think the issue with return is that it allows the result of the constructor to be something other than "this". But hitting this assert seems to imply that we've decided there must be a return statement, otherwise we would not have said "there's a non-this statement". Can you figure out whether that's the case, and how we know?

It would be interesting to make the base constructor try to install a type based on new.target's constructor cache (if it has one). Part of the intent of the subclassing spec is to allow us to optimize this stuff. We should be optimizing the type, inline storage, auxSlot storage, etc. that the base constructor creates.

Reply:
Makes sense. I'll investigate the non-this statement issue.

Reply:
I took a look here and wasn't able to hit the assert I was hitting before with or without setting this flag. Must have ultimately been unrelated to setting the flag here so I put this line back in.

Reply:
Okay. Deferral or non-deferral really shouldn't affect the constructor cache. I'd like to take a look at how constructors use the cache in situations where there is new.target. But that doesn't have to affect this change.